exploration/0002-expression-attributes.md

```diff
@@ -0,0 +1,126 @@
+  - `dir` — An override for the LTR/RTL/auto directionality of the expression.
+
+- Attributes relevant for translators, tools, and other message operations,
+  but with no runtime impact:
+
+  - `example` — A literal value representing
+    what the expression's formatted value will look like.
+  - `note` — A comment on the expression for translators.
+  - `translate` — A boolean `yes`/`no` indicator communicating to translators
+    whether the expression should or should not be localised.
+  - `canCopy`, `canDelete`, `canOverlap`, `canReorder`, etc. — Flags supported by
+    XLIFF 2 inline elements
+
+## Requirements
+
+Attributes should be assignable to un-annotated expressions.
+
+Attributes should be distinct from function options.
+
+Common attributes should be definable by the specification.
+
+Attributes should be user-definable.
+
+Attributes with a runtime impact should not be considered comments.
+
+Attributes should relate to a specific expression.
+
+Multiple attributes should be assignable to a single expression.
+
+Attributes should be assignable to all expressions, not just placeholders.
+
+## Constraints
+
+If supported by new syntax,
+the syntax should be easy to parse by both humans and machines.
+
+If supported by new syntax,
+the reserved/private-use rules will need to be adjusted.
+
+## Proposed Design
+
+Add support for option-like `@key=value` attribute pairs at the end of any expression.
+
+To distinguish expression attributes from options,
+require `@` as a prefix for each attribute asignment.
+Examples: `@translate=yes` and `@locale=$exprLocale`.
+
+Define the meaning and supported values of some expression attributes in the specification,
+including at least `@dir` and `@locale`.
+
+To support later extension of the specified set of attributes while allowing user extensibility,
+require custom attribute names to include a U+002D Hyphen-Minus `-`.
+Examples: `@can-copy=no`, `@note-link=|https://...|`.
+
+Allow expression attributes to influence the formatting context,
+but do not directly pass them to user-defined functions.
+
+## Alternatives Considered
+
+### Do not support expression attributes
+
+If not explicitly defined, less information will be provided to translators.
+
+Function options may be used as a workaround,
+but each implementation and user will end up with different practices.
+
+### Use function options, but with some suggested prefix like `_`
+
+A bit less bad than the previous, but still mixes attributes and options into the same namespace.
+
+At least a no-op function is required for otherwise unannotated expressions.
+
+### Rely on semantic comments
+
+These will be defined within the message resource spec,
+so we introduce a dependence on that.
+
+Referring to specific expressions from outside the message is hard,
+esp. if a similar expression is used in multiple variants.
+
+Comments should not influence the runtime behaviour of a formatter.
```

Comment:
Do you mean:

```suggestion
Comments cannot influence the runtime behavior of a formatter.
```

Reply:
No; I do mean "should not" because this is considering a solution that's based on semantic comments, and then if we do want runtime impact we would indeed have semantic comments influencing runtime behaviour.

Which should not happen, so we need a different solution than semantic comments.